wsnet/conn.go

```diff
@@ -50,7 +46,108 @@ func ConnectEndpoint(baseURL *url.URL, workspace, token string) string {
 	return fmt.Sprintf("%s://%s%s%s%s%s", wsScheme, baseURL.Host, "/api/private/envagent/", workspace, "/connect?session_token=", token)
 }
 
-type conn struct {
+// TURNWebSocketICECandidate returns a valid relay ICEServer that can be used to
+// trigger a TURNWebSocketDialer.
+func TURNProxyICECandidate() webrtc.ICEServer {
+	return webrtc.ICEServer{
+		URLs:           []string{"turn:127.0.0.1:3478?transport=tcp"},
+		Username:       "~magicalusername~",
+		Credential:     "~magicalpassword~",
+		CredentialType: webrtc.ICECredentialTypePassword,
+	}
+}
+
+// TURNWebSocketDialer proxies all TURN traffic through a WebSocket.
+func TURNProxyWebSocket(baseURL *url.URL, token string) proxy.Dialer {
+	return &turnProxyDialer{
+		baseURL: baseURL,
+		token:   token,
+	}
+}
+
+// Proxies all TURN ICEServer traffic through this dialer.
+// References Coder APIs with a specific token.
+type turnProxyDialer struct {
+	baseURL *url.URL
+	token   string
+}
+
+func (t *turnProxyDialer) Dial(network, addr string) (c net.Conn, err error) {
+	headers := http.Header{}
+	headers.Set("Session-Token", t.token)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*15)
+	defer cancel()
+
+	// Copy the baseURL so we can adjust path.
+	url := *t.baseURL
+	url.Scheme = "wss"
+	if url.Scheme == httpScheme {
+		url.Scheme = "ws"
+	}
+	url.Path = "/api/private/turn"
+	conn, resp, err := websocket.Dial(ctx, url.String(), &websocket.DialOptions{
+		HTTPHeader: headers,
+	})
+	if err != nil {
+		if resp != nil {
+			defer resp.Body.Close()
+			return nil, coder.NewHTTPError(resp)
+		}
+		return nil, fmt.Errorf("dial: %w", err)
+	}
+
+	return &turnProxyConn{
+		conn: websocket.NetConn(context.Background(), conn, websocket.MessageBinary),
+	}, nil
+}
+
+// turnProxyConn is a net.Conn wrapper that returns a TCPAddr for the
+// LocalAddr and RemoteAddr functions. pion/ice unsafely checks the
+// types. See: https://github.com/pion/ice/blob/e78f26fb435987420546c70369ade5d713beca39/gather.go#L448
+type turnProxyConn struct {
+	conn net.Conn
+}
```

Comment:
This could just be reimplemented as
```go
type turnProxyConn struct {
    net.Conn
}
```

or even just `type turnProxyConn net.Conn` and override the LocalAddr method

Reply:
Ahh :brain: